content/docs/user-guide/project-structure/dvcyaml-files.md

```diff
@@ -745,6 +745,71 @@ Both individual foreach stages (`train@1`) and groups of foreach stages
 
 </admon>
 
+## `matrix` stages
+
+`matrix` allows you do to define multiple stages based on combinations of
+variables. A `matrix` element accepts one or more variables, each having
+possible list of values to iterate with. For example:
+
+```yaml
+stages:
+  train:
+    matrix:
+      model: [cnn, xgb]
+      feature: [feature1, feature2, feature3]
+    cmd: ./train.py --feature ${item.feature} ${item.model}
+    outs:
+      - ${item.model}.pkl
+```
+
+The values of variables become available in `item` dictionary, and you can
+reference each variable in your stage definition. In above example, you can
+access `item.model` and `item.feature`.
+
+On `dvc repro`, dvc will expand the definition to multiple stages for each
+possible combination of the variables. In above example, dvc will create six
+stages, one for each combination of `model` and`feature`. The name of the stages
+will be generated by appending values of the variables to the stage name after a
+`@` as with foreach. For example, dvc will create following stages:
+
+- _train@cnn-feature1_
+- _train@cnn-feature2_
+- _train@cnn-feature3_
+- _train@xgb-feature1_
+- _train@xgb-feature2_
+- _train@xgb-feature3_
+
+Both individual matrix stages (eg: `train@cnn-feature1`) and group of matrix
+stages (`train`) may be used in commands that accept stage targets.
+
+The values in variables can be simple values such as string, integer, etc and
+composite values such as list, dictionary, etc. For example:
+
+```yaml
+matrix:
+  config:
+    - n_estimators: 150
+      max_depth: 20
+    - n_estimators: 120
+      max_depth: 30
+  labels:
+    - [label1, label2, label3]
+    - [labelX, labelY, labelZ]
+```
+
+When using a list or a dictionary, dvc will generate the name of stages based on
+variable name and id, which is based on index of the value. In above example,
+generated stages may look like `train@labels0-config0`.
+
+Note that templating is supported inside `matrix` and in definition, so you can
+reference things defined in `vars` and from your imported file as usual.
```

Comment:
What do you mean by `your imported file`? Should we also mention that you can reference params files? It would be great to have a more complete example here (show the `vars` section and/or `params.yaml` file).

Reply:
With `imported`,  I meant `vars` loaded from files. But yes, it is confusing. I'll just get rid of it, and replace with an example of `vars` from params.yaml. Others should be left to the readers/users I think.